keanu-project/src/test/java/io/improbable/keanu/util/dot/WriteDotTest.java

```diff
@@ -0,0 +1,166 @@
+        DoubleVertex doubleV2 = new ConstantDoubleVertex(5.43);
+        BoolVertex comparisonV = intV.greaterThan(doubleV2);
+        BoolVertex result2 = boolV.or(comparisonV);
+        result2.setLabel(boolVertexLabel);
+        booleanNet = new BayesianNetwork(result2.getConnectedGraph());
+        WriteDot.outputDot(boolDotFileName, booleanNet);
+        boolDotOutput = new String(Files.readAllBytes(Paths.get(boolDotFileName)), Charset.defaultCharset());
+
+        WriteDot.outputDot(vertexDegree1DotFileName, gaussianV, 1);
+        vertexDegree1DotOutput = new String(Files.readAllBytes(Paths.get(vertexDegree1DotFileName)), Charset.defaultCharset());
+        WriteDot.outputDot(vertexDegree2DotFileName, gaussianV, 2);
+        vertexDegree2DotOutput = new String(Files.readAllBytes(Paths.get(vertexDegree2DotFileName)), Charset.defaultCharset());
+    }
+
+    @Test
+    public void filesGetWrittenAndAreNotOverwritten() throws IOException {
+        String fileName = "dottestfolder/test_file.dot";
+        File outputFile = new File(fileName);
+        FileUtils.deleteDirectory(outputFile.getParentFile());
+
+        WriteDot.outputDot(fileName, gaussianNet);
+        assertTrue(outputFile.exists());
+
+        WriteDot.outputDot(fileName, booleanNet);
+        assertTrue(outputFile.getParentFile().listFiles().length == 2);
+
+        FileUtils.deleteDirectory(outputFile.getParentFile());
+    }
+
+    @Test
+    public void outputContainsAllVertices() {
+        List<Vertex> allGaussianVertices = gaussianNet.getAllVertices();
+
+        for (Vertex v : allGaussianVertices) {
+            assertTrue(gaussianDotOutput.contains(v.getId().hashCode() + ""));
+        }
+
+        List<Vertex> allBoolVertices = booleanNet.getAllVertices();
+
+        for (Vertex v : allBoolVertices) {
+            assertTrue(boolDotOutput.contains(v.getId().hashCode() + ""));
+        }
+    }
+
+    @Test
+    public void outputContainsAllEdges() {
+        List<Vertex> allGaussianVertices = gaussianNet.getAllVertices();
+
+        for (Vertex v : allGaussianVertices) {
+            for (Object vChild : v.getChildren()) {
+                assertTrue(gaussianDotOutput.contains("<" + v.hashCode() + "> -> <" + vChild.hashCode() + ">"));
+            }
+        }
+
+        List<Vertex> allBoolVertices = booleanNet.getAllVertices();
+
+        for (Vertex v : allBoolVertices) {
+            for (Object vChild : v.getChildren()) {
+                assertTrue(boolDotOutput.contains("<" + v.hashCode() + "> -> <" + ((Vertex) vChild).hashCode() + ">"));
+            }
+        }
+    }
+
+    @Test
+    public void outputContainsHyperparameters() throws IOException {
+        assertTrue(gaussianDotOutput.contains("[label=mu]"));
+        assertTrue(gaussianDotOutput.contains("[label=sigma]"));
+    }
+
+    @Test
+    public void outputContainsVertexLabels() {
+        assertTrue(boolDotOutput.contains(boolVertexLabel));
```

Comment:
You're testing a wide range of functionality, which is good, but some of them (like this one) are not very precise: you're checking that 11 chars exist anywhere in a string of almost 200 chars. It would be more natural to test your output against a pre-saved expected file which you can keep in `src/test/resources/`.

You should be able to check that the files are an exact match, as long as you reset `io.improbable.keanu.vertices.VertexId#ID_GENERATOR` to zero in a `@Before` method.

Reply:
Is that the recommended way to unit test functionality with writing outputs? It seems really fragile to any changes. I agree that the current tests could be more exhaustive, but maybe I should improve them instead of going for full output comparison?

Probably better to discuss in person.